core/primitives/src/epoch_manager.rs

```diff
@@ -49,117 +40,975 @@ pub struct EpochConfig {
+            match self {
+                Self::V1(v1) => &v1.last_final_block_hash,
+                Self::V2(v2) => &v2.last_final_block_hash,
+            }
+        }
+
+        #[inline]
+        pub fn prev_hash(&self) -> &CryptoHash {
+            match self {
+                Self::V1(v1) => &v1.prev_hash,
+                Self::V2(v2) => &v2.prev_hash,
+            }
+        }
+
+        #[inline]
+        pub fn epoch_first_block(&self) -> &CryptoHash {
+            match self {
+                Self::V1(v1) => &v1.epoch_first_block,
+                Self::V2(v2) => &v2.epoch_first_block,
+            }
+        }
+
+        #[inline]
+        pub fn epoch_first_block_mut(&mut self) -> &mut CryptoHash {
+            match self {
+                Self::V1(v1) => &mut v1.epoch_first_block,
+                Self::V2(v2) => &mut v2.epoch_first_block,
+            }
+        }
+
+        #[inline]
+        pub fn epoch_id(&self) -> &EpochId {
+            match self {
+                Self::V1(v1) => &v1.epoch_id,
+                Self::V2(v2) => &v2.epoch_id,
+            }
+        }
+
+        #[inline]
+        pub fn epoch_id_mut(&mut self) -> &mut EpochId {
+            match self {
+                Self::V1(v1) => &mut v1.epoch_id,
+                Self::V2(v2) => &mut v2.epoch_id,
+            }
+        }
+
+        #[inline]
+        pub fn chunk_mask(&self) -> &[bool] {
+            match self {
+                Self::V1(v1) => &v1.chunk_mask,
+                Self::V2(v2) => &v2.chunk_mask,
+            }
+        }
+
+        #[inline]
+        pub fn latest_protocol_version(&self) -> &ProtocolVersion {
+            match self {
+                Self::V1(v1) => &v1.latest_protocol_version,
+                Self::V2(v2) => &v2.latest_protocol_version,
+            }
+        }
+
+        #[inline]
+        pub fn slashed(&self) -> &HashMap<AccountId, SlashState> {
+            match self {
+                Self::V1(v1) => &v1.slashed,
+                Self::V2(v2) => &v2.slashed,
+            }
+        }
+
+        #[inline]
+        pub fn slashed_mut(&mut self) -> &mut HashMap<AccountId, SlashState> {
+            match self {
+                Self::V1(v1) => &mut v1.slashed,
+                Self::V2(v2) => &mut v2.slashed,
+            }
+        }
+
+        #[inline]
+        pub fn total_supply(&self) -> &Balance {
+            match self {
+                Self::V1(v1) => &v1.total_supply,
+                Self::V2(v2) => &v2.total_supply,
+            }
+        }
+
+        #[cfg(feature = "protocol_feature_rectify_inflation")]
+        pub fn timestamp_nanosec(&self) -> &u64 {
+            match self {
+                Self::V1(v1) => &v1.timestamp_nanosec,
+                Self::V2(v2) => &v2.timestamp_nanosec,
+            }
+        }
+    }
+
+    #[derive(Default, BorshSerialize, BorshDeserialize, Serialize, Eq, PartialEq, Clone, Debug)]
+    pub struct BlockInfoV2 {
```

Comment:
Can you add a comment on what changed between the versions?

Reply:
Done in https://github.com/near/nearcore/pull/4115/commits/a9f5f9004348c8e8175d8e974303f682d06b3942